diff --git a/tools/Makefile.shared b/tools/Makefile.shared
index 1a24391..774ad2f 100644
--- a/tools/Makefile.shared
+++ b/tools/Makefile.shared
@@ -274,7 +274,7 @@ endif
 
 objinfo_helper$(EXE): objinfo_helper.c ../config/s.h
 	$(BYTECC) $(CCOUT)objinfo_helper$(EXE) $(BYTECCCOMPOPTS) \
-          $(DEF_SYMBOL_PREFIX) $(LIBBFD_INCLUDE) objinfo_helper.c $(LIBBFD_LINK)
+          $(DEF_SYMBOL_PREFIX) $(LIBBFD_INCLUDE) objinfo_helper.c $(LIBBFD_LINK) $(LDFLAGS)
 
 OBJINFO=../compilerlibs/ocamlcommon.cma \
         ../compilerlibs/ocamlbytecomp.cma \
